Issue 152 slf4j migration

ggalmazor opened the pull request:
Closes #152 

This PR adds SLF4J API to the project and Logback binding only for the tests.

All the code that was directly or indirectly using `System.out` and `System.err` has been migrated to SLF4J.

As a consequence, first, deprecation flags have been added to some public artifacts. Then any unused deprecated artifact has been removed. This is also discussed in #52:
- `org.javarosa.core.api.ILogger`
- `org.javarosa.core.io.Std`
- `org.javarosa.core.io.StreamsUtil`
- `org.javarosa.core.services.Logger`
- `org.javarosa.core.util.CodeTimer`
- `org.javarosa.xform.parse.XFormParserReporter`
  - There's a coupling issue with this class. It's used by `FormParserHelper` for logging **and** error handling:

    ```java
    public static ParseResult parse(Path formName) throws IOException {
        XFormParser parser = new XFormParser(new FileReader(formName.toString()));
        final List<String> errorMessages = new ArrayList<>();
        parser.attachReporter(new XFormParserReporter() {
            @Override
            public void warning(String type, String message, String xmlLocation) {
                errorMessages.add(message);
                super.warning(type, message, xmlLocation);
            }
    
            @Override
            public void error(String message) {
                errorMessages.add(message);
                super.error(message);
            }
        });
    
        return new ParseResult(parser.parse(), errorMessages);
    }
    ```

    This is the only special use of `XFormParserReporter` that requires a change on the API. The solution I propose is to add into `XFormParser` a pair of `onWarning` and `onError` methods:

    ```java
    public static ParseResult parse(Path formName) throws IOException {
        XFormParser parser = new XFormParser(new FileReader(formName.toString()));
        final List<String> errorMessages = new ArrayList<>();
        parser.onWarning(new XFormParser.WarningCallback() {
            @Override
            public void accept(String type, String message, String xmlLocation) {
                errorMessages.add(message);
            }
        });
        parser.onError(new XFormParser.ErrorCallback() {
            @Override
            public void accept(String message) {
                errorMessages.add(message);
            }
        });
        return new ParseResult(parser.parse(), errorMessages);
    }
    ```
#### What has been done to verify that this works as intended?
Compiled project and ran automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
I tried to do the minimum amount of changes to do a meaningful migration to slf4j

#### Are there any risks to merging this code? If so, what are they?
Removal of public artifacts has to be explicitly documented on any release note of a JR release that includes this PR

Any error introduced by migrating manual log message variable interpolation to SLF4J's `{}` placeholder syntax. I manually reviewed this three times but there's a risk here of losing information on logs or getting unexpected changes on current log lines.

ggalmazor commented:
I think I broke git history... Maybe I need to recreate it...

dcbriccetti commented on src/org/javarosa/core/io/Std.java:
```diff
@@ -1,32 +0,0 @@
-/**
- * Used to break dependency directly on System.out and System.err.
- */
-public class Std {
```
Did you coordinate with @batkinson, who created this in July?
https://github.com/dcbriccetti/javarosa/commit/f97c27dcf0fbcb39775bde54cd2b4143d5954765

ggalmazor replied:
Not really... It was very useful to guide the refactors, though...

dcbriccetti replied:
This will be a breaking change for him. @lognaturel, should we ask for Brent’s review of this change?

lognaturel replied:
@batkinson wasn't on the TSC call today so I'm guessing he has not had Internet access. :/

How about leaving the `Std` class in, marking it as deprecated and just not using it internally anymore? We could do this with other public classes removed. If we're removing public interfaces we probably should do a major release. By leaving them in we are sure clients can upgrade without changes (even though they may not get the expected logging anymore).

dcbriccetti replied:
Sounds good to me.

ggalmazor replied:
OK, I'll review this option and get back with feedback... Let's hold this PR in the meantime

dcbriccetti commented on .editorconfig:
```diff
@@ -0,0 +1,5 @@
+root = true
```
I’ve been of the view that “meta” files like this, and IDE-specific files, don’t belong in the version control system. Probably because they are not part of the product being developed.

I don’t feel strongly about it, though.

ggalmazor replied:
In my opinion, this is really necessary while there's no better alternative. I'd go even further and add a code style definition for IntelliJ. Ideally, IntelliJ would read our checkstyle rules and configure code style settings to produce complying code but, in the meantime, I think these are very useful.

These things not only help a lot to ease people into start collaborating. They also prevent silly merging conflicts.

I can clearly remember having spent more than one hour monkeying around the code style settings to make the checkstyle rules on Briefcase pass. It was a really frustrating experience that would have been prevented had I have the code style definition already in the repo...

dcbriccetti replied:
OK.

dcbriccetti commented on src/org/javarosa/core/services/ProgramFlow.java:
```diff
@@ -0,0 +1,32 @@
+        final FatalException crashException = new FatalException("unhandled exception in " + thread, e);
+
+        //depending on how the code was invoked, a straight 'throw' won't always reliably crash the app
+        //throwing in a thread should work (at least on our nokias)
```
“at least on our nokias” 
Huh? 
Where did this comment come from?

ggalmazor replied:
¯\_(ツ)_/¯

batkinson replied:
Hah, I do. Javarosa was originally developed for use on Nokia feature phones (I believe part of the original CommCare, the original, not CommCareODK, which came well after) and later the more powerful smart phones. You've found evidence of that fact here. This is likely a *very* old comment.

dcbriccetti commented on src/org/javarosa/core/services/ProgramFlow.java:
```diff
@@ -0,0 +1,32 @@
+public final class ProgramFlow {
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(ProgramFlow.class);
+
+    public static void die(String thread, Exception e) {
```
Maybe add a doc comment, explaining what this is all about—why it’s needed.

ggalmazor replied:
This is actually old code that I've only moved around... I couldn't know. 

I mean, I can interpret that there's some intent of ending "gracefully" normal execution flow, but not much more...

dcbriccetti replied:
Sorry, I thought it was new.

dcbriccetti commented on src/org/javarosa/core/services/PropertyManager.java:
```diff
@@ -104,7 +106,7 @@ public String getSingularProperty(String propertyName) {
         if(retVal == null) {
             //#if debug.output==verbose
-            Std.out.println("Warning: Singular property request failed for property " + propertyName);
+            logger.warn("Singular property request failed for property {}", propertyName);
```
Does this,
`//#if debug.output`
do anything? If not, let’s get rid of them as part of this change.

ggalmazor replied:
I was really puzzled by these comments. It felt like c#'s regions...

dcbriccetti commented on src/org/javarosa/core/services/locale/LocalizationUtils.java:
```diff
@@ -84,7 +87,7 @@ private static void parseAndAdd(OrderedMap<String, String> locale, String line,
                     //Empty Line
                 } else {
-                    Std.out.println("Invalid line (#" + curline + ") read: " + line);
+                    logger.info("Invalid line (#{}) read: {}", curline, line);
```
I assume you weren’t able to test every one of these changes where you added the `{}` placeholders (a change I like), so I’ll just mention the risk of breakage.

ggalmazor replied:
Thanks for noticing :)

I reviewed all the changes trice, but some problem might have slipped... I'll update the issue description to reflect this.

dcbriccetti commented on src/org/javarosa/xpath/expr/XPathExpression.java:
```diff
@@ -76,9 +76,10 @@ public Object pivot (DataInstance model, EvaluationContext evalContext, List<Obj
         for (int i = 0; i < 2 * indent; i++)
-            Std.out.print(" ");
-        Std.out.println(s);
+            padding.append(" ");
```
Here we lack Python’s `' ' * 2 * indent`.

ggalmazor replied:
I'm sure there is a more elegant way to do left padding...

dcbriccetti commented on test/org/javarosa/core/model/data/test/TimeDataTests.java:
```diff
@@ -54,8 +57,6 @@ public static Test suite() {
         return aSuite;
     }
     public static String testMaster (int testID) {
-        //System.out.println("running " + testID);
```
Ah, someone else likes deleting commented-out code. Very nice.

dcbriccetti commented on test/org/javarosa/xform/parse/XFormParserTest.java:
```diff
@@ -79,11 +80,6 @@
     private static final String ORX_2_NAMESPACE_PREFIX = "orx2";
     private static final String ORX_2_NAMESPACE_URI = "http://openrosa.org/xforms";
 
-    @Before
```
Why remove? These code timings are appropriate here.

lognaturel commented:
Agreed that we should wait to hear from @batkinson to merge. I'm guessing he uses `Std` mostly to silence the noise but we should confirm and make sure he's happy with the changes.

dcbriccetti commented:
Do we need to light the Bat(kinson)-Signal? 
https://en.wikipedia.org/wiki/Bat-Signal

@batkinson

ggalmazor commented:
@lognaturel, @dcbriccetti, @batkinson, I've reverted changes on the removed classes and added deprecation annotations and JavaDoc notices wherever I felt there was a clear alternative (in `Logger` it's hard to tell, honestly). 

Please review the latest pair of commits https://github.com/opendatakit/javarosa/pull/259/commits/03d0b84e8e00fc3f0aed10ce8b046770d81425e5 and https://github.com/opendatakit/javarosa/pull/259/commits/0db70cdff7c2fd5d58f0d4010a67d265087fde99

dcbriccetti commented on src/org/javarosa/core/api/ILogger.java:
```diff
@@ -15,38 +15,69 @@
+ * @author Clayton Sims
+ * @deprecated Use {@link org.slf4j.LoggerFactory#getLogger(Class)} instead
  */
+@Deprecated
```
If you deprecate the type, must you also deprecate all the members?

ggalmazor replied:
It's just a pretext to add the `@deprecated` JavaDoc so that anyone ctrl-clicking gets information about an alternative...

dcbriccetti commented on src/org/javarosa/core/util/CodeTimer.java:
```diff
@@ -16,40 +16,59 @@
+ *
+ * @deprecated Use {@link StopWatch} instead
+ */
+@Deprecated
```
Hey, where were you when I wrote this class that turns out to be redundant? :-)

lognaturel commented on src/org/javarosa/core/io/StreamsUtil.java:
```diff
@@ -4,8 +4,11 @@
+import org.slf4j.LoggerFactory;
 
 public class StreamsUtil {
+    private static final Logger logger = LoggerFactory.getLogger(StreamsUtil.class);
```
Is this used?

ggalmazor replied:
I see two `StreamsUtil` classes on `org.javarosa.core.util` and `org.javarosa.core.io` and none are used by JR. It's alarming how they're basically the same classes with slightly different implementations for handling streams. 

I think this classes can be safely removed:
- JR itself doesn't use them.
- No one should depend on a JR's implementation of stream handling functions. There are better and far more battle-tested alternatives like commons-io and anyone relying on these classes should migrate their code a.s.a.p.

lognaturel replied:
👍 How about marking both as deprecated?

ggalmazor replied:
I'm still concerned... I think we should break build if we detect deprecation warnings.

lognaturel commented on src/org/javarosa/xform/parse/XFormParser.java:
```diff
@@ -186,16 +191,15 @@ private static void initProcessingRules() {
-
-                    p.parseControl((IFormElement) parent, e, Constants.CONTROL_INPUT, passedThroughInputAtts,
-                            passedThroughInputAtts);
+                    ));p.parseControl((IFormElement) parent, e, Constants.CONTROL_INPUT,passedThroughInputAtts,
```
Looks like some funny things happened with spacing in the merge!

ggalmazor replied:
I've fixed that, but that class could benefit of static imports a lot. That block could be:
```java
final List<String> passedThroughInputAtts = unmodifiableList(asList("rows","query"));
p.parseControl((IFormElement) parent, e, CONTROL_INPUT, passedThroughInputAtts, passedThroughInputAtts);
```
(less cluttered)

Shall I make that change too?

Also, why is that list `final`? It's used there only to pass it to the `p.parseControl()` method, so, so what if someone reassigns it?

dcbriccetti replied:
I like static imports where they improve readability and shorten code.

Scala has val and var. We use val unless the identifier’s value is allowed to change (which is generally discouraged). val is like Java’s final. The trouble with final in java is that it’s a keyword, and adds noise. So I generally skip it if it’s immediately obvious to the reader that the value doesn’t change, e.g., when its scope is just a few lines. So, I wouldn’t object to removing this final.

lognaturel commented on test/org/javarosa/core/util/test/NumericEncodingTest.java:
```diff
@@ -19,16 +19,18 @@
     public NumericEncodingTest(String name) {
         super(name);
-        System.out.println("Running " + this.getClass().getName() + " test: " + name + "...");
+        logger.info("Running {} test: {}...", this.getClass().getName(), name);
```
I'm wondering whether System.out might be the right interface for a test... I'm not sure! Might be odd to get it in a log file?

ggalmazor replied:
I don't know... I guess that what we should really do is to migrate tests to junit4 and use proper test method names for such feedback messages.

Do you want me to revert the changes on tests?

lognaturel replied:
If you think sending them to the logging makes sense, I'm fine with leaving it. Agreed migrating the tests to junit4 would be ideal. Can that be a quick, surface change or will it require going deeper into each test file?

ggalmazor replied:
I think the migration to junit4 is pretty straightforward thanks to the structure we have followed so far.

One would have to go through the all the 22 test classes and:
- Remove the extension on `TestCase`
- Add a `@Before` annotation on `setUp` and remove the call on `super.setUp()`
- Add a `@Test` annotation on all methods listed on `testMaster()`.
- Remove `testMaster()` and `suite()`
- Remove the class constructor

lognaturel replied:
Sounds great. @ggalmazor you want to just do it (in a PR after this one)?

ggalmazor replied:
OK!

batkinson replied:
So, I have to disagree with @lognaturel's original comment here. Using the logging framework in the tests is the better option IMO. The reason is that it gives you the flexibility to redirect output to a file by default (so you don't blast the console during automated testing), but enable it conditionally while running them during a test debugging session. Really, this is what these frameworks were invented for. Remember those commented `System.out`s you removed? Well, if you log them you never have to leave commented print statements... and... you can enable them at runtime without ever having to recompile the code.

dcbriccetti commented on src/org/javarosa/xform/parse/XFormParser.java:
```diff
@@ -188,36 +191,30 @@ private static void initProcessingRules() {
-                        passedThroughInputAtts,
-                        passedThroughInputAtts
-                    );
+                    List<String> passedThroughInputAtts = unmodifiableList(asList("rows", "query"));
```
I also agree with putting these two string literals side by side. I wonder what my threshold is for wanting them in a column (or columns).

ggalmazor replied:
Set it to 999 :D

dcbriccetti replied:
Right! I forgot who I was talking to. 😀   Mr. “I probably have one of these:”

![dell-u3415w-ultrawide-curved-monitor-review-front-full-2-720x720](https://user-images.githubusercontent.com/55099/37945563-b0fed8fe-3135-11e8-904e-0ec3c78b9a8d.jpg)

ggalmazor replied:
I'm happier since I discovered that shift+mousewheel scrolls horizontally :)

dcbriccetti commented on src/org/javarosa/xform/parse/XFormParser.java:
```diff
@@ -188,36 +191,30 @@ private static void initProcessingRules() {
                 @Override public void handle(XFormParser p, Element e, Object parent) {
-                    p.parseControl((IFormElement) parent, e, Constants.CONTROL_RANGE,
-                        Arrays.asList("start", "end", "step") // Prevent warning about unexpected attributes
+                    p.parseControl((IFormElement) parent, e, CONTROL_RANGE,
```
I like having these `Constants` statically imported.

dcbriccetti commented on src/org/javarosa/xform/parse/XFormParser.java:
```diff
@@ -82,6 +82,9 @@
 
+import static java.util.Arrays.*;
+import static java.util.Collections.*;
+import static org.javarosa.core.model.Constants.*;
```
I like the wildcard import for Constants (because there are so many), but somewhat dislike it here for Arrays and Collections.

ggalmazor replied:
Yeah... static imports on Collections are kind of nice, though...

lognaturel commented on src/org/javarosa/core/io/StreamsUtil.java:
```diff
@@ -5,22 +5,20 @@
 import java.io.OutputStream;
 
+/**
+ * @deprecated Please, don't use anything from this class **at all**. It will removed on the next release
```
I think the deprecation annotation makes the intent clear but if you want to add a comment, perhaps "This class is unused and should remain that way. It will be removed in a future release.

ggalmazor replied:
OK, I'll add the comments on deprecated classes.

This will be my last allegation in favor of breaking the build if there are compilation warnings :) 

Remember that we're already breaking the build for code styling reasons... breaking it for using deprecated stuff makes all the sense to me. Also, what risks do you see, @lognaturel? Maybe I'm missing something...

Right now, there's no warnings at all (I've solved them all). It would be really easy to roll this into the build :)

batkinson commented on src/org/javarosa/xform/parse/XFormParser.java:
```diff
@@ -480,7 +489,7 @@ This will lead to inconsistency between prefixes used in the form definition (bi
     }
 
-    private final Set<String> validElementNames = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+    private final Set<String> validElementNames = unmodifiableSet(new HashSet<>(asList(
```
What's the difference between the final here and the final removed above? This jarred me a little after reading the commit message. In general, I would have split the changes so the commit message explained the *why* and not the action taken. I would have limited the scope of the change to the semantically relevant pieces only if possible.

ggalmazor replied:
Thanks for your suggestions!

Regarding `final` here's my rationale:
- Use `final` on all class fields. Remove it if you really need to reassign and there is no alternative
- Avoid `final` elsewhere

I see value expressing that some member is immutable by hinting it with the `final` keyword but my biggest argument is that it forces you to assign a value on construction, which plays really nice when you want to design your state with value objects. I prefer to work with value objects because it's easier to reason about code with them and they often force you to think about  their validity, emptiness, etc., but that's another discussion.

I wouldn't use `final` elsewhere because they tend to hide bigger problems:
- If you use `final` to protect yourself of a too big/complex method, the real solution is to refactor the method.
- If you use `final` to prevent extending classes to override a method, the design on that class hierarchy is probably flawed.

If your methods are small and your class hierarchies are well designed, adding `final` probably makes code harder to read without adding much value to it, in my opinion.

dcbriccetti replied:
>Avoid final elsewhere

You make good points, @ggalmazor. You will probably agree that in a codebase such as this, with some long and confusing methods, with variables in scope for tens of lines, that it’s nice to know (via `final`) that at least a variable’s value doesn’t change.

ggalmazor replied:
I think they help to understand current code, but if a variable is bound to some meaningful contiguous set of lines, it’s probably better to extract them to a small method.

If the variable is used in many meaningful sets of lines, then extract many small methods and pass the variable as an arg.

If the variable is used with no apparent structure... Then, yeah, `final` brings some good safety restrictions but I think that code would need rework.

batkinson commented on src/org/javarosa/core/api/ILogger.java:
```diff
@@ -15,38 +15,69 @@
- * @date Apr 10, 2009 
  *
+ * @author Clayton Sims
+ * @deprecated Use {@link org.slf4j.LoggerFactory#getLogger(Class)} instead
```
So, I just want to maybe spread a little code-culture here. I would try as much as possible to keep the little edits like you have here in separate commits, clearly explained by their corresponding commit message. It makes reviewing and understanding the scope of commits *way* easier because as soon as you see something that doesn't match the commit message you immediately can question it. Basically, this makes review fatigue immediate and code quality usually suffers as a result. You shouldn't be afraid to do the cleanup, just organize it in incremental commits. Just in case, it's not clear what I'm talking about...

You simultaneously did the following:

- Minor code formatting (whitespace)
- Reorganized imports
- Eliminated the  `@date` javadoc tag 
- Added the `@Deprecated` annotation
- Added the `@deprecated` javadoc tag

You can still make all these changes, but by reorganizing them you can eliminate the need for questions like Dave's here by having the commit message explain exactly *why* you're doing them. It also makes it easy to see your progression by your commit log. Basically, my commit log would look like the above bullets, except I'd reword the last 3 descriptions to explain the reasoning, not the action (which you can see easily using git diff, so the commit message adds little value):

- Eliminated the  `@date` javadoc tag -> Eliminated low-value javadoc noise? (guessing at reason)
- Added the `@Deprecated` annotation -> Annotated recently retired logging api as deprecated
- Added the `@deprecated` javadoc tag -> Provided deprecation suggestions using javadoc tags

You could argue that the last two could be combined, and I agree if kept clean and explainable by:

- Annotated recently deprecated api with javadoc suggestions

ggalmazor replied:
This comment is pure gold. Thanks, @batkinson!

batkinson commented:
So, this looks ok to me, with the caveat that it was difficult to review due due to scope and organization of the changes. I tried to leave enough clues for the reason `Std.err|out` bit was introduced by using the same approach I recommended above. You can get it using the git pickaxe:

```
$ git log -S'Std.err'
commit f97c27dcf0fbcb39775bde54cd2b4143d5954765
Author: Brent Atkinson <brent.atkinson@gmail.com>
Date:   Fri Jul 14 00:13:49 2017 +0100

    Allow client redirection of output sent to stderr and stdout (#151)
```

batkinson commented:
In case it's not 100% clear, by converting the print statements to use the logging framework, the need for the code-based redirection becomes less important. Clients can provide their own logging configuration that redirects the output as needed.

ggalmazor commented:
Thanks, @batkinson for reviewing the PR and for your very helpful insights about how to improve my git skills!

Normally, now I'd rework this PR's git history to start applying all the good stuff from @batkinson's review, but since this PR has already been reviewed by three people, and it's quite big, I'm afraid of introducing errors while doing so and we would have to review everything again.

Is it OK if we leave this PR like it is, or do you think that I should rework the git history?

ggalmazor commented:
@lognaturel, @yanokwa, do you think we can merge this PR?

lognaturel commented:
Yes, thanks, @ggalmazor!
